cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/DataframeCollectionFactory.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.spark.batch;
+
+import io.cdap.cdap.api.data.DatasetContext;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.spark.JavaSparkExecutionContext;
+import io.cdap.cdap.etl.spark.function.FunctionCache;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SQLContext;
+
+/**
+ * Factory that creates a {@link DataframeCollection}
+ */
+public class DataframeCollectionFactory<T> implements BatchCollectionFactory<T> {
+
+  private final Schema schema;
+  private final Dataset<Row> dataFrame;
+
+  /**
+   * Creates dataFrame-based pull result
+   *
+   * @param schema schema of the dataFrame
+   * @param dataFrame result dataFrame
+   */
+  public DataframeCollectionFactory(Schema schema, Dataset<Row> dataFrame) {
+    this.schema = schema;
+    this.dataFrame = dataFrame;
+  }
+
+  @Override
+  public BatchCollection<T> create(JavaSparkExecutionContext sec, JavaSparkContext jsc,
+      SQLContext sqlContext, DatasetContext datasetContext, SparkBatchSinkFactory sinkFactory,
+      FunctionCache.Factory functionCacheFactory) {
+    return (BatchCollection<T>) new DataframeCollection(
```

Comment:
If you make the interface as

```
<V extends BatchCollection<T>> V create(....);
```

Then you don't need to cast here, but have the implementation signature as

```
public DataframeCollection create(...)
```

Reply:
The problem is that we have a bit of a mess with this generics. `DataframeCollection` is `BatchCollection<StructuredRecord>`, not `BatchCollection<T>`. Actually most of the time `BatchCollection` is `BatchCollection<StructuredRecord>`, but sometimes it can be `BatchCollections<RecordInfo<StructuredRecord>>` and it complicates everything. I feel that the RecordInfo thing can be done differently, I even take 1-2 tries to clean it up, but it's tons of changes.

But that's why we need the cast. Essentially we expect T to be StructuredRecord as otherwise DataframeCollection just won't fit